lib/Frontend/CompilerInvocation.cpp

```diff
@@ -973,6 +645,225 @@ StringRef FrontendArgsToOptionsConverter::determineFallbackModuleName() const {
   return llvm::sys::path::stem(nameToStem);
 }
 
+bool FrontendArgsToOptionsConverter::setOutputFilenames() {
+  const std::vector<std::string> unprocessedOutputFlienames =
+  getUnprocessedOutputFilenames();
+  
+  if (!unprocessedOutputFlienames.empty() &&
+      !llvm::sys::fs::is_directory(unprocessedOutputFlienames.back())) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  if (Opts.Inputs.isReadingFromStdin() && unprocessedOutputFlienames.empty()) {
+    Opts.OutputFilenames = unprocessedOutputFlienames;
+    return false;
+  }
+  
+  // No output filename was specified, or an output directory was specified.
+  // Determine the correct output filename.
+  
+  // Note: this should typically only be used when invoking the frontend
+  // directly, as the driver will always pass -o with an appropriate filename
+  // if output is required for the requested action.
+  
+  assert(Opts.actionProducesOutputFromFrontend() ||
+         Opts.OutputFilenames.empty());
```

Comment:
Should this be `^`?

Reply:
Sorry, don't understand the "^". Would you like to clarify?

Reply:
XOR rather than just OR. If the action produces output, there must be output filenames; if the action does not produce output, there must not be output filenames.

Reply:
Thanks for clarifying.
What about "swiftc -frontend -c -primary-file /Users/ungar/main.swift"? An xor would cause the assertion to trip, but today's swiftc works fine.

Reply:
Ah, right, this happens before deriving names. My mistake!